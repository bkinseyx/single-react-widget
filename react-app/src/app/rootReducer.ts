import { combineReducers } from '@reduxjs/toolkit';

import counterReducer from '../features/reduxDemo/reduxDemoSlice';
import profileReducer from '../features/profileCard/profileSlice';
import toDoReducer from '../features/toDoCard/toDoSlice';
import userReducer from '../features/userCard/userSlice';

export const rootReducer = combineReducers({
  reduxDemo: counterReducer,
  profile: profileReducer,
  toDo: toDoReducer,
  user: userReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

export default rootReducer;
