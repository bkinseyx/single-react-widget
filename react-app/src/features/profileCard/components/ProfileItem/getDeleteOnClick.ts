import React from 'react';
import { store } from 'app/store';
import { deleteProfile } from '../../profileSlice';

export const getDeleteOnClick = (profileId?: number) => (
  event: React.MouseEvent,
): void => {
  /* istanbul ignore next */
  if (!profileId) {
    return;
  }
  store.dispatch(deleteProfile(profileId));
  event.stopPropagation();
};
