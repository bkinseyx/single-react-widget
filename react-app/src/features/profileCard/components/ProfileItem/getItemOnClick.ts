import { store } from 'app/store';
import { setActiveProfile } from '../../profileSlice';

export const getItemOnClick = (profileId?: number) => (): void => {
  store.dispatch(setActiveProfile(profileId));
};
