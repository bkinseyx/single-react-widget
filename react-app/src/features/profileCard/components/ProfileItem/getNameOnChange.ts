import { store } from 'app/store';
import { updateProfile } from '../../profileSlice';

export const getNameOnChange = (profileId?: number) => (name: string): void => {
  /* istanbul ignore next */
  if (!profileId) {
    return;
  }
  store.dispatch(
    updateProfile({
      profileId,
      name,
    }),
  );
};
