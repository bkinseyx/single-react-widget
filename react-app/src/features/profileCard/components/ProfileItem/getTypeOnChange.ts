import { store } from 'app/store';
import { updateProfile, ProfileType } from '../../profileSlice';

export const getTypeOnChange = (profileId?: number) => (
  value: string,
): void => {
  /* istanbul ignore next */
  if (!profileId) {
    return;
  }
  store.dispatch(
    updateProfile({
      profileId,
      profileType: value as ProfileType,
    }),
  );
};
