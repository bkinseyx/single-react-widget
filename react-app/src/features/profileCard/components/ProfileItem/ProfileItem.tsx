import React from 'react';
import { useSelector } from 'react-redux';
import styles from './ProfileItem.module.css';
import classNames from 'classnames';

import { Profile } from '../../profileSlice';
import { CustomSelect } from 'app/components/CustomSelect/CustomSelect';
import { CustomTextInput } from 'app/components/CustomTextInput/CustomTextInput';
import { emptyProfile } from '../../helpers/profileTemplate';
import { profileTypesSelector } from '../../selectors/profileTypesSelector';
import { getIsActiveSelector } from '../../selectors/isActiveSelector';
import { getNameOnChange } from './getNameOnChange';
import { getTypeOnChange } from './getTypeOnChange';
import { getItemOnClick } from './getItemOnClick';
import { getDeleteOnClick } from './getDeleteOnClick';

interface ProfileProps {
  profile: Profile;
}

export const ProfileItem: React.FC<ProfileProps> = ({
  profile = emptyProfile,
}) => {
  const profileTypes = useSelector(profileTypesSelector);
  const isActive = useSelector(getIsActiveSelector(profile));

  return (
    <li
      className={classNames('list-group-item', styles.profileItem, {
        active: isActive,
      })}
      onClick={getItemOnClick(profile.profileId)}
    >
      <CustomTextInput
        idPrefix={`name-${profile.profileId}`}
        label={'Name'}
        value={profile.name}
        onChange={getNameOnChange(profile.profileId)}
        autoFocus={isActive}
      />
      <CustomSelect
        idPrefix={`type-${profile.profileId}`}
        label={'Type'}
        value={profile.profileType}
        options={profileTypes}
        onChange={getTypeOnChange(profile.profileId)}
      />
      <div className={styles.buttonWrapper}>
        <button
          className="btn btn-danger"
          onClick={getDeleteOnClick(profile.profileId)}
        >
          Delete
        </button>
      </div>
    </li>
  );
};
