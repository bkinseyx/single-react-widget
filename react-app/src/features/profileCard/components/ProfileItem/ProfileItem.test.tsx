import React from 'react';
import {
  render /*, screen, within*/,
  screen,
  fireEvent,
  RenderResult,
} from '@testing-library/react';
import { Provider } from 'react-redux';

import { store } from 'app/store';
import { ProfileItem } from './ProfileItem';
import { Profile } from 'features/profileCard/profileSlice';

const renderItem = (profile: Profile): RenderResult =>
  render(
    <Provider store={store}>
      <ProfileItem profile={profile} />
    </Provider>,
  );

test('renders ProfileItem', () => {
  const profile = store.getState().profile.profileList[0];
  const { asFragment } = renderItem(profile);
  expect(asFragment()).toMatchSnapshot();
});

test('ProfileItem name field', () => {
  let profile = store.getState().profile.profileList[0];
  const { asFragment, rerender } = renderItem(profile);
  let textInput = screen.getByLabelText('Name');
  expect(textInput).toHaveValue('Ben');
  fireEvent.change(textInput, { target: { value: 'George' } });

  textInput = screen.getByLabelText('Name');
  profile = store.getState().profile.profileList[0];
  expect(profile).toMatchInlineSnapshot(`
    Object {
      "name": "George",
      "profileId": 1,
      "profileType": "admin",
      "showToDoList": true,
    }
  `);
  rerender(
    <Provider store={store}>
      <ProfileItem profile={profile} />
    </Provider>,
  );
  expect(textInput).toHaveValue('George');
  expect(asFragment()).toMatchSnapshot();
});

test('ProfileItem type field', () => {
  let profile = store.getState().profile.profileList[0];
  const { asFragment, rerender } = renderItem(profile);
  let typeSelect = screen.getByLabelText('Type');
  expect(typeSelect).toHaveValue('admin');
  fireEvent.change(typeSelect, { target: { value: 'user' } });

  typeSelect = screen.getByLabelText('Type');
  profile = store.getState().profile.profileList[0];
  expect(profile).toMatchInlineSnapshot(`
    Object {
      "name": "George",
      "profileId": 1,
      "profileType": "user",
      "showToDoList": true,
    }
  `);
  rerender(
    <Provider store={store}>
      <ProfileItem profile={profile} />
    </Provider>,
  );
  expect(typeSelect).toHaveValue('user');
  expect(asFragment()).toMatchSnapshot();
});
