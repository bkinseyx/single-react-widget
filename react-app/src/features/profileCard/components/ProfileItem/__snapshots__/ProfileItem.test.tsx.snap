// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileItem name field 2`] = `
<DocumentFragment>
  <li
    class="list-group-item profileItem"
  >
    <div>
      <label
        for="name-1-input"
      >
        Name
      </label>
      <input
        class="form-control"
        id="name-1-input"
        type="text"
        value="George"
      />
    </div>
    <div>
      <label
        for="type-1-select"
      >
        Type
      </label>
      <select
        class="form-control"
        id="type-1-select"
      >
        <option>
          guest
        </option>
        <option>
          user
        </option>
        <option>
          admin
        </option>
      </select>
    </div>
    <div
      class="buttonWrapper"
    >
      <button
        class="btn btn-danger"
      >
        Delete
      </button>
    </div>
  </li>
</DocumentFragment>
`;

exports[`ProfileItem type field 2`] = `
<DocumentFragment>
  <li
    class="list-group-item profileItem"
  >
    <div>
      <label
        for="name-1-input"
      >
        Name
      </label>
      <input
        class="form-control"
        id="name-1-input"
        type="text"
        value="George"
      />
    </div>
    <div>
      <label
        for="type-1-select"
      >
        Type
      </label>
      <select
        class="form-control"
        id="type-1-select"
      >
        <option>
          guest
        </option>
        <option>
          user
        </option>
        <option>
          admin
        </option>
      </select>
    </div>
    <div
      class="buttonWrapper"
    >
      <button
        class="btn btn-danger"
      >
        Delete
      </button>
    </div>
  </li>
</DocumentFragment>
`;

exports[`renders ProfileItem 1`] = `
<DocumentFragment>
  <li
    class="list-group-item profileItem"
  >
    <div>
      <label
        for="name-1-input"
      >
        Name
      </label>
      <input
        class="form-control"
        id="name-1-input"
        type="text"
        value="Ben"
      />
    </div>
    <div>
      <label
        for="type-1-select"
      >
        Type
      </label>
      <select
        class="form-control"
        id="type-1-select"
      >
        <option>
          guest
        </option>
        <option>
          user
        </option>
        <option>
          admin
        </option>
      </select>
    </div>
    <div
      class="buttonWrapper"
    >
      <button
        class="btn btn-danger"
      >
        Delete
      </button>
    </div>
  </li>
</DocumentFragment>
`;
